common/js/src/messaging/single-message-based-channel-unittest.js

```diff
@@ -0,0 +1,312 @@
+});
+
+// Verify that the message channel passes the messages received from the other
+// side to the correct service while also preserving the relative order.
+goog.exportSymbol('testSingleMessageBasedChannelReceiving', function() {
+  var testCasePromiseResolver = goog.Promise.withResolver();
+  var propertyReplacer = new goog.testing.PropertyReplacer;
+
+  setupSendMessageMock(testCasePromiseResolver, propertyReplacer);
+  setupSendMessagePingResponding();
+
+  /** @type {?} */ chrome.runtime.sendMessage;
+
+  chrome.runtime.sendMessage.$replay();
+
+  function onChannelEstablished() {
+    var receivedMessages = [];
+
+    for (let testMessage of TEST_MESSAGES) {
+      globalChannel.registerService(
+          testMessage.type,
+          function(messageData) {
+            receivedMessages.push(new TypedMessage(
+                testMessage.type, messageData));
+          },
+          true);
+    }
+
+    for (let testMessage of TEST_MESSAGES) {
+      globalChannel.deliverMessage(testMessage.makeMessage());
+    }
+
+    chrome.runtime.sendMessage.$verify();
+    testCasePromiseResolver.resolve();
+    globalChannel.dispose();
+  }
+
+  function onChannelDisposed() {
+    testCasePromiseResolver.reject();
+  }
+
+  globalChannel = new GSC.SingleMessageBasedChannel(
+      EXTENSION_ID, onChannelEstablished);
+  globalChannel.addOnDisposeCallback(onChannelDisposed);
+
+  return testCasePromiseResolver.promise;
+});
+
+// Verify that the message channel is disposed of when the underlying pinger
+// stops receiving responses (but first the channel needs to be established).
+goog.exportSymbol('testSingleMessageBasedChannelDisposing', function() {
+  var testCasePromiseResolver = goog.Promise.withResolver();
+  var propertyReplacer = new goog.testing.PropertyReplacer;
+
+  setupSendMessageMock(testCasePromiseResolver, propertyReplacer);
+  propertyReplacer.set(Pinger, 'TIMEOUT_MILLISECONDS', 400);
+  propertyReplacer.set(Pinger, 'INTERVAL_MILLISECONDS', 200);
+  /** @type {?} */ chrome.runtime.sendMessage;
+
+  var firstCallSeen = false;
+  chrome.runtime.sendMessage(isStringMockMatcher, isPingMessageMockMatcher)
+      .$atLeastOnce().$does(
+          function(message) {
+            if (!firstCallSeen) {
+              firstCallSeen = true;
+              globalChannel.deliverMessage(pingResponseMessage);
+            }
+          });
+
+  chrome.runtime.sendMessage.$replay();
+
+  var channelEstablished = false;
+  function onChannelEstablished() {
+    channelEstablished = true;
+  }
+
+  function delayFunction() {
+    assert('Channel wasn\'t successfully established prior to failure',
+           channelEstablished);
+    assert(globalChannel.isDisposed());
+
+    assertThrows(function() {
+      globalChannel.send('foo', {});
+    });
+
+    chrome.runtime.sendMessage.$verify();
+    testCasePromiseResolver.resolve();
+    globalChannel.dispose();
+  }
+
+  new goog.async.Delay(delayFunction, 800).start();
```

Comment:
Indeed. Didn't even realise that because I was thinking too much in terms of a delayed check, thanks.